metronome/checkpointing/service/src/io/iohk/metronome/checkpointing/service/storage/LedgerStorage.scala

```diff
@@ -0,0 +1,42 @@
+package io.iohk.metronome.checkpointing.service.storage
+
+import io.iohk.metronome.checkpointing.models.Ledger
+import io.iohk.metronome.storage.{KVRingBuffer, KVCollection, KVStore}
+import scodec.Codec
+
+/** Storing the committed and executed checkpoint ledger.
+  *
+  * Strictly speaking the application only needs the committed state,
+  * since it has been signed by the federation and we know it's not
+  * going to be rolled back. Uncommitted state can be kept in memory.
+  *
+  * However we want to support other nodes catching up by:
+  * 1. requesting the latest Commit Q.C., then
+  * 2. requesting the block the Commit Q.C. points at, then
+  * 3. requesting the ledger state the header points at.
+  *
+  * We have to allow some time before we get rid of historical state,
+  * so that it doesn't disappear between step 2 and 3, resulting in
+  * nodes trying and trying to catch up but always missing the beat.
+  *
+  * Therefore we keep a collection of the last N ledgers in a ring buffer.
+  */
+class LedgerStorage[N](
+    ledgerColl: KVCollection[N, Ledger.Hash, Ledger],
```

Comment:
Would it make sense to use `Block.Hash` as the key instead? It is a 1:1 relationship and we could obtain the parent block's state directly rather than first obtaining the block to read its `postStateHash`. `Ledger.Hash` would then only be used for validation. Of course that only gives us 1 hop in the block path, so I'm not sure if it's really useful.

Reply:
Great question, I wanted to do this too in the robot example. The reason I didn't is because I figured that if we did that, then there would be no way to validate whether the state returned in the [getState](https://github.com/input-output-hk/metronome/pull/42/files#diff-fcfbaee6667bbaa3ca47408851c4d5b631b40a7143915eb9ed020d267328e3f8R177-R187) response was valid. Now that I'm thinking about it again, it's still possible if we have the block.

So, I want the state at block B, at the moment I retrieve it by `getState(block.postStateHash)` and verify that `response.state.hash == postStateHash`. In the world where state is stored by block key, it would be just `getState(block)` and verifying that `response.state.hash == block.postStateHash`. 

I think I just didn't think it through enough :confused: 

Another drawback could be that if the state doesn't change at all, we still insert it again into the store.

Reply:
OTOH I just realised that my code assumes that it can delete a state from the collection if it's in the oldest key position, it doesn't consider that the interim positions may refer to it as well :disappointed: 
Storing by block hash wouldn't have that problem.

Reply:
What I really wanted was not to store state hash in the block, that's what confused me. If not wasn't there, then it would really be impossible to validate a state we fetched.

Reply:
Fixed it by refcounting. It's up to you what key you want to use. You are right that it could save 1 query in the robot example each time I'm projecting the state from the last executed block hash.